src/vocabulary/mod.rs

```diff
@@ -136,143 +179,122 @@ impl Vocabulary {
     }
 }
 
-impl Vocabulary {
-    /// Inserts a token to the vocabulary with the specified identifier.
-    pub fn insert(mut self, token: impl Into<Token>, id: TokenId) -> Vocabulary {
-        self.insert_in_place(token, id);
-        self
-    }
-
-    /// Extends the vocabulary with tokens and their identifiers.
-    pub fn extend<T: Into<Token>, I: IntoIterator<Item = TokenId>>(
-        mut self,
-        tokens_and_ids: impl IntoIterator<Item = (T, I)>,
-    ) -> Vocabulary {
-        self.extend_in_place(tokens_and_ids);
-        self
-    }
-}
-
-impl Vocabulary {
-    /// Inserts a token to the vocabulary with the specified identifier, in place.
-    pub fn insert_in_place(&mut self, token: impl Into<Token>, id: TokenId) {
-        // TODO: return error if eos token id is inserted
-        let token = token.into();
-        self.tokens.entry(token).or_default().push(id);
-    }
-
-    /// Extends the vocabulary with tokens and their identifiers, in place.
-    pub fn extend_in_place<T: Into<Token>, I: IntoIterator<Item = TokenId>>(
-        &mut self,
-        tokens_and_ids: impl IntoIterator<Item = (T, I)>,
-    ) {
-        for (token, ids) in tokens_and_ids.into_iter() {
-            let token = token.into();
-            self.tokens.entry(token).or_default().extend(ids);
-        }
-    }
-}
-
-impl std::ops::Deref for Vocabulary {
-    type Target = FxHashMap<Token, Vec<TokenId>>;
-
-    fn deref(&self) -> &FxHashMap<Token, Vec<TokenId>> {
-        &self.tokens
-    }
-}
-
 impl std::fmt::Display for Vocabulary {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        for (index, (token, token_ids)) in self.iter().enumerate() {
-            if index != (self.len() - 1) {
-                writeln!(f, "{:?} -> {:?}", token, token_ids)?;
-            } else {
-                write!(f, "{:?} -> {:?}", token, token_ids)?;
-            }
+        writeln!(
+            f,
+            "Vocabulary object with eos_token_id={:?} and the following tokens to token_ids:",
+            self.eos_token_id
+        )?;
+        for (token, token_ids) in self.tokens.iter() {
+            writeln!(
+                f,
+                "{:?} -> {:?}",
+                token
+                    .iter()
+                    .map(|b| format!("0x{:02X}", b))
+                    .collect::<Vec<_>>(),
+                token_ids
+            )?;
         }
         Ok(())
     }
 }
 
-impl From<FxHashMap<Token, Vec<TokenId>>> for Vocabulary {
-    fn from(tokens: FxHashMap<Token, Vec<TokenId>>) -> Vocabulary {
-        Vocabulary {
-            eos_token_id: None,
-            tokens,
+impl TryFrom<(TokenId, HashMap<Token, Vec<TokenId>>)> for Vocabulary {
```

Comment:
These `TryFrom` implementations are on very complex types. I think it might be better to make them regular functions:
```rust
Vocabulary::try_from((eos_token_id, token_map))
```
vs
```rust
Vocabulary::from_token_map(eos_token_id, token_map)
```

Reply:
Well, type won't become simpler if we'll add a custom function. 

**But** instead of a very known trait we'd force users to _search through all_ our documentation to find this custom conversion method. Users often have to answer this question: how we can convert something into something and they know **exactly** in which section of the docs to go to see the options. And we're also getting automatic docs generations for this.

So, the rule of thumb is always to use a idiomatic and ergonomic approach first, unless it's not possible for some reason, but this reason will be very well defined.

![image](https://github.com/user-attachments/assets/969e1837-41e1-40ce-afd8-aa2df4201bd2)

![image](https://github.com/user-attachments/assets/f71a8262-66c6-45b2-aff6-cfbf1bad033f)

Reply:
It's just not very easy to understand what `u32` or `HashMap<String, Vec<u32>, FxBuildHasher>` is. With an explicit method, you can name the arguments appropriately.

Here, it's not mentioned that first `u32` is the `eos_token_id`, so it can be confusing IMO.

Reply:
Yes, but it's still not clear that the first `TokenId` is the EOS token id. In any case, this was just a suggestion, it's okay to keep these :+1:

Reply:
I get your point, but yeah, even with this downside I think it's still preferable to keep them as traits outside of the custom interface.